stdlib/src/builtin/string.mojo

```diff
@@ -211,6 +211,198 @@ fn ascii(value: String) -> String:
 # ===----------------------------------------------------------------------=== #
 
 
+@always_inline
+fn _stol(str_ref: StringRef, base: Int = 10) raises -> (Int, StringRef):
+    """Implementation if `stol` for StringRef inputs.
+
+    Please see its docstring for details.
+    """
+    if (base != 0) and (base < 2 or base > 36):
+        raise Error("Base must be >= 2 and <= 36, or 0.")
+
+    if not str_ref:
+        raise Error("Cannot convert empty string to integer.")
+
+    var result: Int = 0
+    var real_base: Int
+    var start: Int = 0
+    var is_negative: Bool = False
+    var str_len = len(str_ref)
+    var buff = str_ref.unsafe_ptr()
+
+    start, is_negative = _trim_and_handle_sign(str_ref, str_len)
+
+    if start == str_len or not _is_valid_digit(int(buff[start]), base):
+        return 0, str_ref
+
+    start = _handle_base_prefix(start, str_ref, str_len, base)
+
+    if base == 0:
+        if start == (str_len - 1):
+            real_base = 10
+        elif str_ref[start] == "0" and start + 1 < str_len:
+            var second_digit = str_ref[start + 1]
+            if second_digit == "b" or second_digit == "B":
+                real_base = 2
+                start += 1  # Move past the '0', but not the 'b'
+            elif second_digit == "o" or second_digit == "O":
+                real_base = 8
+                start += 1
+            elif second_digit == "x" or second_digit == "X":
+                real_base = 16
+                start += 1
+            else:
+                real_base = 10
+
+            # Check if the character after the prefix is valid
+            if real_base != 10:
+                if start + 1 < str_len and _is_valid_digit(
+                    int(buff[start + 1]), real_base
+                ):
+                    start += 1  # Move past the prefix character
+                else:
+                    # Invalid prefix or digit after prefix
+                    return 0, StringRef(
+                        str_ref.unsafe_ptr() + start, str_len - start
+                    )
+        else:
+            real_base = 10
+    else:
+        real_base = base
+
+    var ord_num_max: Int
+    var ord_letter_max = (-1, -1)
+    alias ord_0 = ord("0")
+    var ord_letter_min = (ord("a"), ord("A"))
+    alias ord_underscore = ord("_")
+
+    if real_base <= 10:
+        ord_num_max = ord(str(real_base - 1))
+    else:
+        ord_num_max = ord("9")
+        ord_letter_max = (
+            ord("a") + (real_base - 11),
+            ord("A") + (real_base - 11),
+        )
+
+    var was_last_digit_underscore = True
```

Comment:
```suggestion
    var was_last_digit_underscore = False
```
here you could have e.g. "0b_1_0" or "0x_1_0" or "0o_1_0" wrongly breaking out in the first iteration